### PR TITLE
Implemented Probeless method

### PR DESCRIPTION
**Pull request by nadirdurrani**
Implementation of Probless method for neuron analysis

### CODE REVIEW

File: neurox/interpretation/probeless.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
+"""Module for Probeless method
+This module extracts neuron ranking for an attribute or for the entire property set .
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: I think better to use the language "tag" or "label" instead of "attribute" and "property set" to maintain consistency in the toolkit; I'm sure we break this elsewhere but lets not add to the chaos :D

File: neurox/interpretation/probeless.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+.. see also::
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: add new line before this, `seealso` must be a single word without space

File: neurox/interpretation/probeless.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        `Antverg, Omer and Belinkov, Yonatan "On The Pitfalls Of Analyzing Idividual Neurons in Language Models." In Proceedings of the 10th International Conference on Learning Representations (ICLR).
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: End this line with "`_" (accent + underscore)

File: neurox/interpretation/probeless.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+"""
+import numpy as np
+import torch
+import itertools
+import sys
+import torch.nn as nn
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: I think this is unused, remove

File: neurox/interpretation/probeless.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+from torch.autograd import Variable
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: remove unused import

File: neurox/interpretation/probeless.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+from . import metrics
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: metrics and utils also unused it seems?

File: neurox/interpretation/probeless.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+from . import utils
+from . import ablation
+
+
+def get_mean_vectors(X_t, Y_t):
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: prefix all "internal functions" that you don't expect an end user to use with an underscore; here function name will be `_get_mean_vectors`

File: neurox/interpretation/probeless.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+def overall_ranking(X_a, y_t, NUM_LAYERS, layer=-1):
+    
+    layer_X_train = get_layer_vector(X_a, y_t, NUM_LAYERS, layer)
+    avg_embeddings, average_embeddings_by_label = get_mean_vectors(layer_X_train, y_t)
+    ranking = get_overall_ranking(avg_embeddings)
+    
+    return ranking
+
+def tag_wise_ranking(X_a, y_t , tag, NUM_LAYERS, layer=-1):
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: Do you think it might be worth unifying or having some similarities between neuron ranking methods here and linear_probe? So renaming this function as `get_top_neurons`

This way, someone using LCA will call:
`linear_probe.get_top_neurons`

and using probeless will call:
`probless.get_top_neurons`

makes it much easier for end user

Comment (fdalvi) [reply neurox/interpretation/probeless.py]: this may be more applicable to ranking_for_all_tags function I think

File: neurox/interpretation/probeless.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+from . import ablation
+
+
+def get_mean_vectors(X_t, Y_t):
+    
+    embeddings_by_labels = {}
+    average_embeddings_by_label = {}
+    rankings_per_label = {}
+
+    for words in range(Y_t.size):
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: Append `_idx` to index variables so its easy to understand the code that follows; i.e. `words` -> `word_idx`

File: neurox/interpretation/probeless.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+def tag_wise_ranking(X_a, y_t , tag, NUM_LAYERS, layer=-1):
+
+    layer_X_train = get_layer_vector(X_a, y_t, NUM_LAYERS, layer)
+    avg_embeddings, average_embeddings_by_label = get_mean_vectors(layer_X_train, y_t)
+    summation, ranking = get_tag_wise_ranking(average_embeddings_by_label, tag)
+
+    return ranking
+
+def ranking_for_all_tags(X_a, y_t, NUM_LAYERS, layer=-1):
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: What is `X_a`, `y_t` here? Ideally we should accept the same arguments as linear_probe sibling functions, i.e. the output of utils.create_tensors

**Review by fdalvi — CHANGES_REQUESTED** [review]
Please add comments for all functions (most importantly for user facing functions) - see linear_probe for examples and syntax

Thanks a lot for going through the process!

File: neurox/interpretation/probeless.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
+"""Module for Probeless method
+
+This module extracts neuron ranking for a label/tag (e.g Verbs) or for the entire property set (e.g Part of speech).
+
+.. seealso::
+        `Antverg, Omer and Belinkov, Yonatan "On The Pitfalls Of Analyzing Idividual Neurons in Language Models." In Proceedings of the 10th International Conference on Learning Representations (ICLR).`_
+"""
+import numpy as np
+import torch
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: torch is unused

File: neurox/interpretation/probeless.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+    """
+
+    layer_X_t = _get_layer_vector(X_t, y_t, NUM_LAYERS, layer)
+    avg_embeddings, average_embeddings_by_label = _get_mean_vectors(layer_X_t, y_t)
+    summation, ranking = _get_tag_wise_ranking(average_embeddings_by_label, tag)
+
+    return ranking
+
+def get_top_neurons_for_all_tags(X_t, y_t, NUM_LAYERS, layer=-1):
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: change "top_neurons" to "ordering", since "top neurons" means a set of neurons and not a ranking

File: neurox/interpretation/probeless.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    for c, qzz in mean_vectors_by_label.items():
+        
+        if (tag != c):
+            summation = np.add(summation, np.abs(np.subtract(qz,qzz)))
+    
+    ranking = np.argsort(summation)[::-1].tolist()
+    return summation, ranking 
+
+
+def _get_layer_vector(X_t, y_t, NUM_LAYERS, layer=-1):
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: remove all references to layer manipulation, it should be done outside of neuron discovery (use can use ablation.filter before calling get_neuron_ordering

File: neurox/interpretation/probeless.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+    """
+
+    layer_X_t = _get_layer_vector(X_t, y_t, NUM_LAYERS, layer)
+    avg_embeddings, average_embeddings_by_label = _get_mean_vectors(layer_X_t, y_t)
+    summation, ranking = _get_tag_wise_ranking(average_embeddings_by_label, tag)
+
+    return ranking
+
+def get_top_neurons_for_all_tags(X_t, y_t, NUM_LAYERS, layer=-1):
+
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: all these use facing functions should take label2idx and return results with class names instead of "class #"

File: neurox/interpretation/probeless.py [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+        summation, ranking = _get_tag_wise_ranking(average_embeddings_by_label, c)
+        overall = np.add(overall, summation)
+        ranking_per_tag [c] = ranking
+
+    overall_ranking = np.argsort(overall)[::-1].tolist()
+
+    return overall_ranking, ranking_per_tag
+
+
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: remove all these beautiful empty lines

File: neurox/interpretation/probeless.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+def get_top_neurons_for_all_tags(X_t, y_t, NUM_LAYERS, layer=-1):
+
+
+    """
+    Returns a dictionary of tags along with top neurons for each tag
+    Returns a list of overall ranking
+
+    Parameters
+    ----------
+    X_t : numpy.ndarray
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: If X_t is referening to train, rename to `X_train`, if it can be any tensor in general, rename to `X`; same for `y_t`

File: neurox/interpretation/probeless.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+    Numpy Matrix of size [``NUM_TOKENS`` x ``NUM_NEURONS``]. Usually the 
+    output of ``interpretation.utils.create_tensors``
+    y_t : numpy.ndarray
+    Numpy Vector of size [``NUM_TOKENS``] with class labels for each input
+    token. Usually the output of ``interpretation.utils.create_tensors``.
+    NUM_LAYERS : int
+    Total number of layers in the network
+    layer : int, optional
+    layer number for which ranking needs to be extracted, -1 for entire network
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py]: In documentation, all the descriptions should be indented one more level; so:
```
y_t : numpy.ndarray
description
```

should be 
```
y_t : numpy.ndarray
    description
```

File: neurox/interpretation/probeless.py [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+    tag : int 
+    tag for which rankings are extracted
+    NUM_LAYERS : int
+    Total number of layers in the network
+    layer : int, optional
+    layer number for which ranking needs to be extracted, -1 for entire network
+
+    Returns
+    -------
+    ranking : list
```
Comment (fdalvi) [inline neurox/interpretation/probeless.py:101]: Please add description (list of size X ordered in descending order of importance etc etc)

**Review by fdalvi — COMMENTED** [review]
Changes look very nice; small push and this should be good to go!
